server/src/main/java/org/apache/druid/server/ClientQuerySegmentWalker.java

```diff
@@ -840,6 +846,7 @@ private static <T, QueryType extends Query<T>> Optional<DataSource> materializeR
                                 + "from the query context and/or the server config."
                             );
       } else {
+        resultSequence.set(results);
         return Optional.empty();
```

Comment:
why do we want to fallback - instead of telling the user that this have failed; and he might want to try the row based limiting?

I also feel like catching `Exception` is a bit broad; as it could catch `NullPointerException` or `AIOOBE` and such as well

Reply:
> as it could catch NullPointerException or AIOOBE and such as well

That won't happen. This is because `startedAccumulating` indicates that the query processing has began. NullPtrException and AIOOBE will only happen if we have started processing the subquery results, which would mean that `startedAccumulating = true`. And if that's the case, for any random `Exception`, we rethrow it.

Reply:
* that boolean is not `true` but this will be catched in [this block](https://github.com/apache/druid/pull/16800/files/350661e41fc6cc5ec2d0a86bf3b46749e54bfc0a#diff-073196043cf52e8c14ca92d8df79bc9e17b20e6df3fb60ec0a97f78b5d896595R783-R795)
* which also contains `toolChest#resultsAsFrames` which has a few implementations  - any of which could throw an `Exception` and will be routed to a retry.
* isn't a possible problematic case would be if the signature contains a column starting with `___druid` ; which produces an `IAE`? 
   * `TimeseriesQQTC#resultsAsFrames` -> `FrameWriters.makeColumnBasedFrameWriterFactory` -> `disallowedFieldNames.isEmpty()` -> `IAE` ?


could you please answer my original question:
> why do we want to fallback - instead of telling the user that this have failed; and he might want to try the row based limiting?

Reply:
> why do we want to fallback - instead of telling the user that this have failed; and he might want to try the row based limiting?

Copied from one of my responses: 

Fallback is mostly for when the types aren't known. I agree that it is a performance hit, but at the time this feature was added, the signature informed by the tool chest didn't need to have a type. Scan queries only had knowledge of the column names (and not types), group by/time series... etc. toolchests could return null for the aggregator's dimensions. The fallback was present for these cases, where it's easy to detect the failure relatively early in the whole subquery processing flow. Fallback meant that transitioning from row -> byte based limit was simple. There's an undocumented parameter that treated these null types as JSON types, but that had logical flaws of its own iirc.

Reply:
your answer suggests that the situation have changed - couldn't it be changed to fallback early and don't even try?

Reply:
> your answer suggests that the situation have changed

Yes, the fallback cases are mostly known and preventable. 

> couldn't it be changed to fallback early and don't even try?

I think it would require a change in the tool chests since there isn't a method currently to get the row signature while specifying the finalization. Well, essentially `resultArraySignature` would deem the aggregator's signature as finalization = UNKNOWN even though we want the finalization = YES.